File: cpp/3_longest_substring_without_repeating_characters.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
		int left = 0, right = 0;
		int ans = 0;
		unordered_set<char> set;
		while (right < s.length()) {
			if (set.count(s[right]) > 0) {
				set.erase(s[left]);
				left++;
			} else {
				ans = max(ans, right - left + 1);
				set.insert(s[right]);
				right++;
			}
		}
		return ans;
    }
};

File: python/3_longest_substring_without_repeating_characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """

        seen = set()
        l = 0
        r = 0
        ret = 0
        while (r < len(s)):
            while (s[r] in seen):
                seen.remove(s[l])
                l += 1
            seen.add(s[r])
            r += 1
            ret = max(ret, r - l)
        return ret
